core/src/program/mod.rs

```diff
@@ -126,15 +126,39 @@ impl CodeBlockTable {
 #[derive(Debug, Clone, Default, PartialEq, Eq)]
 pub struct Kernel(Vec<Digest>);
 
+// TODO: This value should be capped to `u8::MAX`.
+pub const MAX_KERNEL_PROCEDURES: usize = u16::MAX as usize;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum KernelError {
+    TooManyProcedures(usize),
+}
+
+impl Display for KernelError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            KernelError::TooManyProcedures(count) => write!(
+                f,
+                "Kernel can have at most {} procedures, received {}",
+                MAX_KERNEL_PROCEDURES, count
+            ),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for KernelError {}
+
 impl Kernel {
     /// Returns a new [Kernel] instantiated with the specified procedure hashes.
-    pub fn new(proc_hashes: &[Digest]) -> Self {
-        // make sure procedure roots are ordered consistently
-        let mut hash_map: BTreeMap<[u8; 32], Digest> = BTreeMap::new();
-        proc_hashes.iter().cloned().for_each(|r| {
-            hash_map.insert(r.into(), r);
-        });
-        Self(hash_map.values().copied().collect())
+    pub fn new(proc_hashes: &[Digest]) -> Result<Self, KernelError> {
+        if proc_hashes.len() > MAX_KERNEL_PROCEDURES {
+            Err(KernelError::TooManyProcedures(proc_hashes.len()))
+        } else {
+            let mut hashes = proc_hashes.to_vec();
+            hashes.sort(); // ensure consistent order
```

Comment:
I hope the original code was not being used to remove duplicates. If it was, then this needs a `hashes.dedup()`

Reply:
I think it was - but the better approach is to return an error if a duplicate is detected. Let's add this as another variant to `KernelError`.

Reply:
added check